### PR TITLE
refactor: major refactor for memory saving

### PR DESCRIPTION
**Pull request by zhuzilin**
This PR tries to diminish the memory occupancy of megatron:
1. use gloo for most gather_object and all_gather_object
~2. add an warmup to the model initialization~ The warmup doesn't solve the issue completely.
3. only initialize one model and make ref or old_actor as cpu state_dict

### CODE REVIEW

File: slime/backends/megatron_utils/model.py [hunk 1]
```diff
@@ -110,10 +109,99 @@ def setup_model_and_optimizer(
         skip_load_to_model_and_opt=False,
     )
 
+    warmup(args, model, optimizer)
+
     print(f"Loaded checkpoint at iteration {iteration}")
     return model, optimizer, opt_param_scheduler, iteration
 
 
+def warmup(args, model, optimizer):
```
Comment (fzyzcjy) [inline slime/backends/megatron_utils/model.py]: curious whether we can put all training steps under mem_saver.region()

Comment (zhuzilin) [reply slime/backends/megatron_utils/model.py]: yeah, I believe this is a better idea than `warmup`. I will remove the warmup here and try to put the whole training procedure under `mem_saver.region()`.

Comment (fzyzcjy) [reply slime/backends/megatron_utils/model.py]: sure! if you find any errors just ping me (e.g. I expect some tensors may be used for cross-gpu communication and you need torch_memory_saver.disable() on it to avoid crashes).
